core/drivers/versal_net_nvm.c

```diff
@@ -0,0 +1,1060 @@
+
+	if (!p)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	val = SHIFT_U32(p->reg_init_dis & 0x3, 30) |
+	      SHIFT_U32(p->boot_env_wr_lk & 0x1, 28) |
+	      SHIFT_U32(p->sec_lock_dbg_dis & 0x3, 21) |
+	      SHIFT_U32(p->sec_dbg_dis & 0x3, 19) |
+	      SHIFT_U32(p->user_key1_wr_lk & 0x1, 15) |
+	      SHIFT_U32(p->user_key1_crc_lk & 0x1, 14) |
+	      SHIFT_U32(p->user_key0_wr_lk & 0x1, 13) |
+	      SHIFT_U32(p->user_key0_crc_lk & 0x1, 12) |
+	      SHIFT_U32(p->aes_wr_lk & 0x1, 11) |
+	      SHIFT_U32(p->aes_crc_lk & 0x3, 9) |
+	      SHIFT_U32(p->ppk2_wr_lk & 0x1, 8) |
+	      SHIFT_U32(p->ppk1_wr_lk & 0x1, 7) |
+	      SHIFT_U32(p->ppk0_wr_lk & 0x1, 6) |
+	      SHIFT_U32(p->jtag_dis & 0x1, 2) |
+	      SHIFT_U32(p->jtag_err_out_dis & 0x1, 1) |
+	      (p->aes_dis & 0x1);
+
+	return do_write_efuses_value(EFUSE_WRITE_SEC_CTRL_BITS, val);
+}
+
+TEE_Result versal_efuse_write_misc(struct versal_efuse_misc_ctrl_bits *p)
+{
+	uint32_t val = 0;
+
+	if (!p)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	val = SHIFT_U32(p->glitch_det_halt_boot_en & 0x3, 30) |
+	      SHIFT_U32(p->glitch_det_rom_monitor_en & 0x1, 29) |
+	      SHIFT_U32(p->halt_boot_error & 0x3, 21) |
+	      SHIFT_U32(p->halt_boot_env & 0x3, 19) |
+	      SHIFT_U32(p->crypto_kat_en & 0x1, 15) |
+	      SHIFT_U32(p->lbist_en & 0x1, 14) |
+	      SHIFT_U32(p->safety_mission_en & 0x1, 8) |
+	      SHIFT_U32(p->ppk2_invalid & 0x3, 6) |
+	      SHIFT_U32(p->ppk1_invalid & 0x3, 4) |
+	      SHIFT_U32(p->ppk0_invalid & 0x3, 2);
+
+	return do_write_efuses_value(EFUSE_WRITE_MISC_CTRL_BITS, val);
+}
+
+TEE_Result
+versal_efuse_write_glitch_cfg(struct versal_efuse_glitch_cfg_bits *p)
+{
+	uint32_t val = 0;
+
+	if (!p)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!p->prgm_glitch)
+		return TEE_SUCCESS;
+
+	val = SHIFT_U32(p->glitch_det_wr_lk & 0x1, 31) | p->glitch_det_trim;
+
+	return do_write_efuses_value(EFUSE_WRITE_GLITCH_CONFIG, val);
+}
+
+TEE_Result
+versal_efuse_write_boot_env(struct versal_efuse_boot_env_ctrl_bits *p)
+{
+	uint32_t val = 0;
+
+	if (!p)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	val = SHIFT_U32(p->sysmon_temp_en & 0x1, 21) |
+	      SHIFT_U32(p->sysmon_volt_en & 0x1, 20) |
+	      SHIFT_U32(p->sysmon_temp_hot & 0x3, 17) |
+	      SHIFT_U32(p->sysmon_volt_pmc & 0x3, 12) |
+	      SHIFT_U32(p->sysmon_volt_pslp & 0x3, 10) |
+	      SHIFT_U32(p->sysmon_volt_soc & 0x3, 8) |
+	      (p->sysmon_temp_cold & 0x2);
+
+	return do_write_efuses_value(EFUSE_WRITE_BOOT_ENV_CTRL_BITS, val);
+}
+
+TEE_Result versal_efuse_write_sec_misc1(struct versal_efuse_sec_misc1_bits *p)
+{
+	uint32_t val = 0;
+
+	if (!p)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	val = SHIFT_U32(p->lpd_mbist_en & 0x7, 10) |
+	      SHIFT_U32(p->pmc_mbist_en & 0x7, 7) |
+	      SHIFT_U32(p->lpd_noc_sc_en & 0x7, 4) |
+	      SHIFT_U32(p->sysmon_volt_mon_en & 0x3, 2) |
+	      (p->sysmon_temp_mon_en & 0x3);
+
+	return do_write_efuses_value(EFUSE_WRITE_MISC1_CTRL_BITS, val);
+}
+
+TEE_Result versal_efuse_write_offchip_ids(uint32_t id)
```

Comment:
this function is incorrect, there are total 8 offchip_revoke_id, and we use the api to update values for certain id, the parameters is lacking of the values going into that offchip id.
Please refer to the implementation in versal_nvm.c

Reply:
Hi @wangjudyw,

Thanks for your feedback. This implementation is a direct mapping of the API offered by the xilnvm service:

https://github.com/Xilinx/embeddedsw/blob/86a272bd9cd412bf1e5214f6098aefff301e58f0/lib/sw_services/xilnvm/src/versal_net/server/xnvm_efuse_cdohandler.c#L229

As you can see, it only expects an uint32_t for the ID to be written in the fuses (and a flag that is set by default by the `do_write_efuses_value()` helper function). Could you elaborate what you mean?

Thanks!